## DRILL
Return to the sorting wiki page and pick an algorithm we haven't covered here (you probably want to pick one of the simpler ones, but it's up to you. Implement it in Python below and see how it compares in sorting our short and long lists. You should be able to easily find guides on how to implement any of the algorithms. Can you figure out why it runs faster or slower than our other sorting algorithms?

Some good sorts to try are:
* Heap Sort
* Selection Sort
* QuickSort

In [22]:
import time
import random

# Set seed.
random.seed(a=100)

# Create our default list.
short_list = list(random.sample(range(1000000), 10))
long_list = list(random.sample(range(1000000), 1000))

### Heapsort

In [23]:
from heapq import heappush, heappop
def heapsort(iterable):
    h = []
    for value in iterable:
        heappush(h, value)
    return [heappop(h) for i in range(len(h))]

In [24]:
start_time = time.time()

heapsort(short_list)

print("--- %s seconds ---" % (time.time() - start_time))

--- 9.608268737792969e-05 seconds ---


In [25]:
start_time = time.time()

heapsort(long_list)

print("--- %s seconds ---" % (time.time() - start_time))

--- 0.0007810592651367188 seconds ---


### Selection Sort

In [32]:
def selection_sort(a_list):
    for i in range(len(a_list)):
        # Find the minimum element in remaining
        minPosition = i
        
        for j in range(i+1, len(a_list)):
            if a_list[minPosition] > a_list[j]:
                minPosition = j
        # Swap the found minimum element with minPosition
        temp = a_list[i]
        a_list[i] = a_list[minPosition]
        a_list[minPosition] = temp
        
    return a_list

In [33]:
start_time = time.time()

selection_sort(short_list)

print("--- %s seconds ---" % (time.time() - start_time))

--- 9.703636169433594e-05 seconds ---


In [34]:
start_time = time.time()

selection_sort(long_list)

print("--- %s seconds ---" % (time.time() - start_time))

--- 0.05864381790161133 seconds ---


### Quicksort

In [29]:
# This function takes last element as pivot, places 
# the pivot element at its correct position in sorted 
# array, and places all smaller (smaller than pivot) 
# to left of pivot and all greater elements to right 
# of pivot 
        
def quicksort(array):
    
    less = []
    equal = []
    greater = []

    if len(array) > 1:
        pivot = array[0]
        for x in array:
            if x < pivot:
                less.append(x)
            elif x == pivot:
                equal.append(x)
            elif x > pivot:
                greater.append(x)
        return quicksort(less)+equal+quicksort(greater)
# When there is only one element in the array, just return the array
    else:  
        return array

In [30]:
start_time = time.time()

quicksort(short_list)

print("--- %s seconds ---" % (time.time() - start_time))

--- 0.00011086463928222656 seconds ---


In [31]:
start_time = time.time()

quicksort(long_list)

print("--- %s seconds ---" % (time.time() - start_time))

--- 0.10862517356872559 seconds ---


### Conclusion

It looks like selection sort took a while sorting our long list with 8.2 seconds. Heapsort did fine with sorting long list. But overall, the sort method buili in python is still the fasted (and by far the easiest method to use). 

Heapsort:

* short: 0.00009608268737792969 seconds
* long: 0.0007810592651367188 seconds

Selection Sort:

* short: 0.00009894371032714844 seconds
* long: 0.054344892501831055 seconds

Quicksort:

* short: 0.00011086463928222656 seconds
* long: 0.10862517356872559 seconds
* extra long: recursion error!

Heapsort is the clear winner here...selection sort is not far off for the short list, but there is a marked difference for the long list. Qicksort is, well...definitely not quick, compared to the other 2 at least.
